internal/client/consensus/grandpa/until_imported.go

```diff
@@ -0,0 +1,336 @@
+					_, ok := ui.pending[item.TargetHash]
+					if !ok {
+						ui.pending[item.TargetHash] = pendingEntry[H, N, Blocked]{
+							BlockNumber: item.TargetNumber,
+							LastLog:     time.Now(),
+							Wait:        make([]blockUntilImported[H, N, Blocked], 0),
+						}
+					}
+					p := ui.pending[item.TargetHash]
+					p.Wait = append(ui.pending[item.TargetHash].Wait, item.Wait)
+					ui.pending[item.TargetHash] = p
+				}
+			case ready[Blocked]:
+				ui.ready.PushBack(dwr.Ready)
+			}
+		default:
+			break incoming
+		}
+	}
+
+imports:
+	for {
+		select {
+		case notification, ok := <-ui.importNotifications:
+			if !ok {
+				return true, nil, nil
+			}
+			// new block imported. queue up all messages tied to that hash.
+			if _, ok := ui.pending[notification.Hash]; ok {
+				p := ui.pending[notification.Hash]
+				delete(ui.pending, notification.Hash)
+				canonNumber := notification.Header.Number()
+				readyMessages := make([]Blocked, 0)
+				for _, m := range p.Wait {
+					blocked := m.WaitCompleted(canonNumber)
+					if blocked != nil {
+						readyMessages = append(readyMessages, *blocked)
+					}
+				}
+				for _, m := range readyMessages {
+					ui.ready.PushBack(m)
+				}
+			}
+		default:
+			break imports
+		}
+	}
+
+	var updateInterval bool
+	select {
+	case <-ui.checkPending:
+		updateInterval = true
+	default:
+	}
+
+	if updateInterval {
+		knownKeys := make([]HashNumber[H, N], 0)
+		for blockHash, e := range ui.pending {
+
+			number, err := ui.statusCheck.Number(blockHash)
+			if err != nil {
+				return true, nil, err
+			}
+			if number != nil {
+				knownKeys = append(knownKeys, HashNumber[H, N]{Hash: blockHash, Number: *number})
+			} else {
+				nextLog := e.LastLog.Add(logPendingInterval)
+				if time.Now().After(nextLog) {
+					logger.Debugf(
+						"Waiting to import block %s before %d %s messages can be imported. "+
+							"Requesting network sync service to retrieve block from. Possible fork?",
+						blockHash, len(e.Wait), ui.identifier)
+
+					// NOTE: when sending an empty vec of peers the underlying should make a best effort to sync the
+					// block from any peers it knows about.
+					ui.blockSyncRequester.SetSyncForkRequest(
+						nil,
+						blockHash,
+						e.BlockNumber,
+					)
+					e.LastLog = nextLog
+					ui.pending[blockHash] = e
+				}
+			}
+		}
+
+		for _, hn := range knownKeys {
+			if _, ok := ui.pending[hn.Hash]; ok {
+				p := ui.pending[hn.Hash]
+				delete(ui.pending, hn.Hash)
+				readyMessages := make([]Blocked, 0)
+				for _, m := range p.Wait {
+					blocked := m.WaitCompleted(hn.Number)
+					if blocked != nil {
+						readyMessages = append(readyMessages, *blocked)
+					}
+				}
+				for _, m := range readyMessages {
+					ui.ready.PushBack(m)
+				}
+			}
+		}
+	}
+
+	if ui.ready.Len() > 0 {
+		ready := ui.ready.PopFront()
+		return true, &ready, nil
+	}
+
+	return false, nil, nil
+}
+
+func warnAuthorityWrongTarget[H runtime.Hash](hash H, id primitives.AuthorityID) {
```

Comment:
is there a reason this needs to be its own function?

Reply:
We use the same error twice so we use a function.